src/xdg.rs

```diff
@@ -221,42 +208,125 @@ impl Icon {
+        for data_dir in data_dirs {
+            // Find whether each dir has icons, in the only theme we support for now.
+            //
+            // In the future, we might want to parse the index.theme of the theme we want to
+            // load, for instance to handle the proper inheritance hierarchy.
+            let mut icons_dir = data_dir.to_owned();
+            icons_dir.push("icons");
+            icons_dir.push(theme_name);
+
+            for dir_entry in fs::read_dir(icons_dir).unwrap() {
+                let dir_entry = dir_entry.unwrap();
+                let file_name = dir_entry.file_name();
+                let file_name = match file_name.to_str() {
+                    Some(file_name) => file_name,
+                    None => continue,
+                };
 
-        // Check all paths for icons.
-        //
-        // Since the `ICON_PATHS` is in reverse order of our priority, we can just
-        // insert every new icon into `icons` and it will correctly return the
-        // closest match.
-        for (path, extension) in data_dirs
-            .iter()
-            .flat_map(|base| ICON_PATHS.iter().map(|(path, ext)| (base.join(path), ext)))
-        {
-            let mut read_dir = fs::read_dir(path).ok();
-            let entries = read_dir.iter_mut().flatten().flatten();
-            let files =
-                entries.filter(|e| e.file_type().map_or(false, |e| e.is_file() || e.is_symlink()));
-
-            // Iterate over all files in the directory.
-            for file in files {
-                let file_name = file.file_name().to_string_lossy().to_string();
-
-                // Store icon paths with the correct extension.
-                let name = file_name.rsplit_once('.').filter(|(_, ext)| ext == extension);
-                if let Some((name, _)) = name {
-                    let _ = icons.insert(name.to_owned(), file.path());
+                // Each directory in an icon theme has the same layout, so let’s hardcode it.
+                let bucket = if file_name == "scalable" {
+                    Bucket::Scalable
+                } else if file_name == "symbolic" {
+                    Bucket::Symbolic
+                } else if let [width, height] = file_name.split('x').collect::<Vec<_>>()[..] {
+                    match (u32::from_str(width), u32::from_str(height)) {
+                        (Ok(width), Ok(height)) if width == height => Bucket::Exact(width),
+                        _ => continue,
+                    }
+                } else {
+                    continue;
+                };
+
+                let mut dir_path = dir_entry.path();
+                dir_path.push("apps");
+                let read_dir = match fs::read_dir(dir_path) {
+                    Ok(read_dir) => read_dir,
+                    Err(_) => continue,
+                };
+                for file in read_dir {
+                    let file = file.unwrap();
```

Comment:
Should probably flatten here rather than unwrap. Currently this can crash.

Reply:
I went with a `try_continue!()` macro to remove all of the `.unwrap()` calls.